### PR TITLE
soundwire: sdca: parse function type and controls

### PR DESCRIPTION
**Pull request by plbossart**
SDCA devices are based on independent 'Functions', which are exposed in platform firmware using the DisCo 2.1 specification.

This patch adds a mechanism to parse the functions and extract basic information (adr and function type).

Example log on a MeteorLake device (which already shows problems with ACPI information...)

[   13.069728] soundwire sdw:0:0:025d:0713:01: found invalid SDCA function type 4 ADR 1, skipped
[   13.069743] soundwire sdw:0:0:025d:0713:01: found SDCA function type 6 ADR 3
[   13.070166] soundwire sdw:0:1:025d:1316:01: found SDCA function type 1 ADR 4
[   13.070750] soundwire sdw:0:2:025d:1316:01: found SDCA function type 1 ADR 4

### CODE REVIEW

**Comment by plbossart** [comment]
v2: all the SDCA information is moved to a sound/soc/sdca library, and the information is parsed before the driver probe (if present).

**Comment by bardliao** [comment]
Thanks @plbossart Looks like a good start for SDCA driver.

**Comment by plbossart** [comment]
@shumingfan would you happen to know what this Function 4 might be. I wonder if the function topology uses an older version of the SDCA/DisCo spec?

**Comment by plbossart** [comment]
logs with the last version

````
[   51.528012] rt1316-sdca sdw:0:1:025d:1316:01: found SDCA function AF04 type 1 ADR 4
[   51.528023] rt1316-sdca sdw:0:1:025d:1316:01: AF04: SDCA function has topology-features mask 0xb00
[   51.528033] acpi device:28: find_sdca_entities: AF04: found entity 0x1 type 0x2 label IT 21
[   51.528038] acpi device:28: find_sdca_entities: AF04: found entity 0x2 type 0x21 label UDMPU 21
[   51.528043] acpi device:28: find_sdca_entities: AF04: found entity 0x3 type 0x7 label FU 21
[   51.528047] acpi device:28: find_sdca_entities: AF04: found entity 0x33 type 0x11 label PDE 22
[   51.528052] acpi device:28: find_sdca_entities: AF04: found entity 0x22 type 0xa label XU 22
[   51.528056] acpi device:28: find_sdca_entities: AF04: found entity 0x29 type 0x24 label SAPU 29
[   51.528061] acpi device:28: find_sdca_entities: AF04: found entity 0x54 type 0x21 label UDMPU 23
[   51.528065] acpi device:28: find_sdca_entities: AF04: found entity 0x55 type 0x7 label FU 23
[   51.528070] acpi device:28: find_sdca_entities: AF04: found entity 0x31 type 0x11 label PDE 23
[   51.528074] acpi device:28: find_sdca_entities: AF04: found entity 0x56 type 0x3 label OT 23
[   51.528078] acpi device:28: find_sdca_entities: AF04: found entity 0x51 type 0x2 label IT 29
[   51.528082] acpi device:28: find_sdca_entities: AF04: found entity 0x34 type 0x11 label PDE 24
[   51.528086] acpi device:28: find_sdca_entities: AF04: found entity 0x24 type 0xa label XU 24
[   51.528091] acpi device:28: find_sdca_entities: AF04: found entity 0x52 type 0x3 label OT 24
[   51.528095] acpi device:28: find_sdca_entities: AF04: found entity 0x53 type 0xb label CS 24
[   51.528100] acpi device:28: find_sdca_entities: AF04: found entity 0x21 type 0xb label CS 21
[   51.528104] acpi device:28: find_sdca_entities: AF04: found entity 0x40 type 0x30 label TG 23
[   51.528638] rt1316-sdca sdw:0:2:025d:1316:01: found SDCA function AF04 type 1 ADR 4
[   51.528643] rt1316-sdca sdw:0:2:025d:1316:01: AF04: SDCA function has topology-features mask 0xb00
[   51.528651] acpi device:2a: find_sdca_entities: AF04: found entity 0x1 type 0x2 label IT 21
[   51.528655] acpi device:2a: find_sdca_entities: AF04: found entity 0x2 type 0x21 label UDMPU 21
[   51.528659] acpi device:2a: find_sdca_entities: AF04: found entity 0x3 type 0x7 label FU 21
[   51.528665] acpi device:2a: find_sdca_entities: AF04: found entity 0x33 type 0x11 label PDE 22
[   51.528668] acpi device:2a: find_sdca_entities: AF04: found entity 0x22 type 0xa label XU 22
[   51.528673] acpi device:2a: find_sdca_entities: AF04: found entity 0x29 type 0x24 label SAPU 29
[   51.528677] acpi device:2a: find_sdca_entities: AF04: found entity 0x54 type 0x21 label UDMPU 23
[   51.528681] acpi device:2a: find_sdca_entities: AF04: found entity 0x55 type 0x7 label FU 23
[   51.528685] acpi device:2a: find_sdca_entities: AF04: found entity 0x31 type 0x11 label PDE 23
[   51.528689] acpi device:2a: find_sdca_entities: AF04: found entity 0x56 type 0x3 label OT 23
[   51.528693] acpi device:2a: find_sdca_entities: AF04: found entity 0x51 type 0x2 label IT 29
[   51.528698] acpi device:2a: find_sdca_entities: AF04: found entity 0x34 type 0x11 label PDE 24
[   51.528712] acpi device:2a: find_sdca_entities: AF04: found entity 0x24 type 0xa label XU 24
[   51.528718] acpi device:2a: find_sdca_entities: AF04: found entity 0x52 type 0x3 label OT 24
[   51.528722] acpi device:2a: find_sdca_entities: AF04: found entity 0x53 type 0xb label CS 24
[   51.528726] acpi device:2a: find_sdca_entities: AF04: found entity 0x21 type 0xb label CS 21
[   51.528730] acpi device:2a: find_sdca_entities: AF04: found entity 0x40 type 0x30 label TG 23
[   51.529357] rt712-sdca sdw:0:0:025d:0713:01: found invalid SDCA function AF01 type 4 ADR 1, skipped
[   51.529374] rt712-sdca sdw:0:0:025d:0713:01: found SDCA function AF03 type 6 ADR 3
[   51.529383] acpi device:26: find_sdca_entities: AF03: found entity 0x1 type 0x31 label (null)
````

Now we need to figure out something useful with all this.

**Comment by shumingfan** [comment]
> @shumingfan would you happen to know what this Function 4 might be. I wonder if the function topology uses an older version of the SDCA/DisCo spec?

I will try to ask the hardware team about this. If there are any results, I will give feedback.

**Comment by vijendarmukunda** [comment]
Thanks @plbossart for pushing the SDCA driver patches. We will also try to validate sdca tables parsing  on AMD platform.

**Comment by shumingfan** [comment]
> I will try to ask the hardware team about this. If there are any results, I will give feedback.

The hardware engineer thinks reporting the function type 4 is a mistake.
The rt713 only needs the function type 6 actually. The rt1713 should report the mic function.

**Comment by plbossart** [comment]
> The hardware engineer thinks reporting the function type 4 is a mistake. The rt713 only needs the function type 6 actually. The rt1713 should report the mic function.

Right, that's what I thought @shumingfan. In this device there's no RT1713 (local mic disabled), so I can't check further.

File: sound/soc/sdca/sdca_functions.c [hunk 1]
```diff
@@ -30,5 +30,85 @@ int sdca_alloc(void **context,
...
+	struct sdca_data *sdca = data;
+	struct device *parent = sdca->parent;
+	struct fwnode_handle *control5; /* used to identify topology type */
+	bool valid_function_found = false;
+	u32 function_topology = 0;
+	acpi_status status;
+	u64 addr;
+
+	status = acpi_evaluate_integer(adev->handle,
+				       METHOD_NAME__ADR, NULL, &addr);
```
Comment (charleskeepax) [inline sound/soc/sdca/sdca_functions.c:313]: Should probably acpi_get_local_address

File: sound/soc/sdca/sdca_functions.c [hunk 2]
```diff
@@ -191,6 +192,18 @@ static int find_sdca_function(struct acpi_device *adev, void *data)
...
+	fwnode_property_read_u32(function_node, "mipi-sdca-function-busy-max-delay",
+				 &max_busy_delay);
+	sdca->functions[sdca->function_count].function_busy_max_delay_us =
+		max_busy_delay;
+	if (max_busy_delay > 0) {
+		dev_err(parent,
+			"%pfwP: unsupported non-zero delay %d us for Function_Busy\n",
+			function_node, max_busy_delay);
+		dev_err(parent,
+			"please report this on https://github.com/thesofproject/linux/issues\n");
```
Comment (charleskeepax) [inline sound/soc/sdca/sdca_functions.c:380]: Well you say there is no hardware that uses this feature, but cs42l43 and likely its successors make heavy use of it if they are used in SDCA mode. Not that support needs to be added right now, but just making you aware.

File: sound/soc/sdca/sdca_functions.c [hunk 3]
```diff
@@ -119,6 +119,140 @@ static int find_sdca_entities(struct device *dev,
...
+	 *  in this Function. Entity0 is an illegal connection since
+	 *  it is used for Function-level configurations
+	 */
+	if (input_pin_list & BIT(0)) {
+		dev_err(dev, "%s: %pfwP: entity_id %#x has invalid input_pin 0\n",
+			__func__, function_node, entity->id);
+		return -EINVAL;
+	}
+
+	for(pin = 1; pin < 64; pin++) {
```
Comment (charleskeepax) [inline sound/soc/sdca/sdca_functions.c:165]: Probably slightly nice to use BITS_PER_TYPE(input_pin_list)

File: sound/soc/sdca/sdca_functions.c [hunk 4]
```diff
@@ -119,6 +119,140 @@ static int find_sdca_entities(struct device *dev,
...
+		struct fwnode_handle *connected_node;
+		struct sdca_entity *connected_entity;
+		const char *connected_label;
+		char pin_property[40];
```
Comment (charleskeepax) [inline sound/soc/sdca/sdca_functions.c:169]: I would be tempted to add some sort of define for this max length, lot of max 40 byte strings used in the code.

File: sound/soc/sdca/sdca_functions.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sdca/sdca_functions.c:313]: Erm, no. that function is wrong. _ADR is a 64-bit value, but the code assumes it's 32-bits.

I vaguely remember we had to fix some of the ACPI subsystem to make _ADR a true 64 bit value, this function has not been updated apparently.

Actually now I do remember, and it's even in git:

````
commit ca6f998cf9a259fe6019d44768064e0cfe8a925b
Author: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
Date:   Wed May 1 07:53:22 2019 -0500

    ACPI: bus: change _ADR representation to 64 bits
````

So this code would need to be updated to u64.

````c
int acpi_get_local_address(acpi_handle handle, u32 *addr)
{
	unsigned long long adr;
	acpi_status status;

	status = acpi_evaluate_integer(handle, METHOD_NAME__ADR, NULL, &adr);
	if (ACPI_FAILURE(status))
		return -ENODATA;

	*addr = (u32)adr;
	return 0;
}
````

Gah.

**Comment by plbossart** [comment]
closing for now, will re-add the functionality when basic parsing is implemented with PR #4995

**Pull request closed without merging**